### PR TITLE
The hitscanening.

### PR DESCRIPTION
**Pull request by Stalkeros2**
## About The Pull Request
Makes all bullet and laser projectiles hitscan.
This also includes plasma cutter because I felt like it and I really wanna see it shoot those three blue beams "like Dead Space" (but this time with no insane damage; just aesthetics).

## How This Contributes To The Skyrat Roleplay Experience
Moving away from OOC-player-skill-based pretty-much-gaming-centric /tg/ combat in favor of everyone being equal just like Colt intended.

## Proof of Testing

<details>
<summary>Screenshots/Videos</summary>
I shoota da gunna.

image

I shoota da other gunna; tracer's **right there**, but it's barely visible because it's the marksman tracer sprite and let's be honest why would it even be more visible (this is me making up an excuse to make this placeholder sprite permanent; it still, actually, does what it should do tbh).
image

And the third one just for clarity.
image

</details>

## Changelog

:cl: Stalkeros
refactor: Makes (a lot) of projectiles hitscan.
/:cl:

### CODE REVIEW

**Comment by Stalkeros2** [comment]
I hope the refactor tag's right to use here; I forgot what it means tbh.

**Comment by gonenoculer5** [comment]
Im pretty sure we tried this before and it didnt go well.

**Comment by Stalkeros2** [comment]
> Im pretty sure we tried this before and it didnt go well.

Iirc it was only locked to a single gun (basic lasgun; pretty cool unique feature tbh) and it was reverted for the reason of "it not being like /tg/".
Which is, you know, **haha lol lmao** in itself.

**Comment by theselfish** [comment]
I really wish I had something to say here.

Admitting to making combat less skillful is just uhh...
I'm unsure what to say other than *why*.

Why is removing the ability to have skills in combat good in any way?

**Comment by W4rd3nn** [comment]
Been saying this for, like, ever. Nearly all combat-oriented codebases have either incredibly fast projectiles or hitscan weapons. Simply because they are just better.

Not to mention it will interact better with our insane TIDI. Guns should be lethal without the capability to step around the bullet because you can see it coming strictly due to the time dilation and/or lag that's occurring.

**Comment by axietheaxolotl** [comment]
based PR

**Comment by gonenoculer5** [comment]
Depends on what style of hitscan it is. Could be a good change or a terrible one depending on how it was implemented.

**Comment by softcerv** [comment]
I feel like by making them hitscan, you remove tradeoff of being able to quickly and easily recharge at the cost of only firing projectiles.

File: modular_skyrat/modules/gunsgalore/code/projectile.dm [hunk 1]
```diff
@@ -92,3 +92,18 @@
...
+
+/obj/projectile/beam
+	tracer_type = /obj/effect/projectile/tracer/laser
+	hitscan = TRUE
+	hitscan_light_intensity = 0.75
+
+/obj/projectile/plasma
+	impact_effect_type = /obj/effect/temp_visual/impact_effect/blue_laser
+	hitscan = TRUE
+	hitscan_light_intensity = 0.75
```
Comment (softcerv) [inline modular_skyrat/modules/gunsgalore/code/projectile.dm:110]: How is this going to impact mining?

Comment (Stalkeros2) [reply modular_skyrat/modules/gunsgalore/code/projectile.dm:110]: Not much really, considering the speed at which the normal plasma cutter projectile's usually going (difference here being: VERY fast it's basically instant vs instant);
Maybe miners will start hitting each other more frequently due to the lack of telegraphing but hey not really that much of a problem here.

**Comment by Iamgoofball** [comment]
hitscans suck for everyone, they suck for people who lag, they aren't fun and nobody likes hitscan except the asshole with the good connection dual wielding the laser pistols

i'm so tired of having to explain this every year when someone gets the genius idea that we need to try hitscan again with no changes because "it's different this time" and "okay it didn't work the last 5 times but it'll work this time!!!"

normally i'd do the big writeup on why hitscans suck but gamers just keep wearing me down by repeatedly PRing hitscan because "hitscan cool" at every single downstream known to man every year

i need to remove the hitscan code from upstream so people who don't understand why hitscan sucks on a fundamental level for SS13 quit turning it back on every year, its just a yearly problem with any ss13 codebase that has support for hitscan projectiles

**Comment by Stalkeros2** [comment]
> I feel like by making them hitscan, you remove tradeoff of being able to quickly and easily recharge at the cost of only firing projectiles.

**Sure do!**

**Comment by Iamgoofball** [comment]
``Moving away from OOC-player-skill-based pretty-much-gaming-centric /tg/ combat in favor of everyone being equal just like Colt intended.``

this is, in fact, not going to make anyone equal, and is just going to make people really really hate ranged combat in laggy scenarios

**Comment by Iamgoofball** [comment]
here are the relevant PRs and forum threads that have the discourse around hitscan in SS13 combat and why it sucks from professional SS13 maintainers, developers, and community members
https://github.com/tgstation/tgstation/pull/20142 feat: oranges predicting the future
https://github.com/tgstation/tgstation/pull/57080 we ran hitscan projectiles as an actual test, and players fuckin' did not like it

if this isn't enough, here's some feedback from players who tried this exact hitscan system under non-laggy conditions where the server wasn't perpetually running 160+ player rounds, everyone who posted in this thread actually played with hitscan lasers live:
https://tgstation13.org/phpBB/viewtopic.php?f=10&t=28611
note that the prevailing feedback is "unfair for the receiving end" and "extremely overpowered"

**Comment by Iamgoofball** [comment]
also, this is a minor one because literally who the fuck is aimbotting on goddamn skyrat of all places, but anyone rocking aim assist of any kind is going to become an unstoppable murder machine with hitscan projectiles

**Comment by press-mark** [comment]
skyrat gun fetishism example #90151548

**Comment by Iamgoofball** [comment]
@coldud13 don't post unproductive comments

**Comment by W4rd3nn** [comment]
Well some things. The tracer on bullets kinda stinky and bad. You can barely tell where they're going which is realistic but also lame.

Also they don't seem to properly generate impact effects when striking objects other than people.

Additionally, this makes bow and arrows hitscan as well.

**Comment by Stalkeros2** [comment]
> Well some things. The tracer on bullets kinda stinky and bad. You can barely tell where they're going which is realistic but also lame.

Really is a placeholder until I get a proper sprite. (I'll probably honestly just un-mostly-transparent the current one.)

> Also they don't seem to properly generate impact effects when striking objects other than people.

They do; check the plasma cutter screenshot. Unless you're talking about the bullet example specifically, in which case, I don't think I can do much without getting a proper sprite.

> Additionally, this makes bow and arrows hitscan as well.

Fixed in the next commit.

**Comment by YakumoChen** [comment]
This is really just going to make ranged combat the only viable option.

Laser guns were very overbearing before when we made them hitscan.

Not to mention this tips the scales in favour of security in fights so hard that you might as well and lie down and give up if you fight more than one at once.

All even fights become a contest of "who decided to click first / has lower ping after the ritual handshake of Combat Starts Now"

Takes all the guns with gimmicky traits and cranks the power of them to 11. 

Hitscan buckshot

Hitscan ebows

Hitscan accelerator lasers

**Comment by Stalkeros2** [comment]
> This is really just going to make ranged combat the only viable option.

With the sheer amount of ranged buffs and Skyrat's very ranged-centric combat, to imply it wasn't already is kinda funny.

> Laser guns were very overbearing before when we made them hitscan.

image
 
> Not to mention this tips the scales in favour of security in fights so hard that you might as well and lie down and give up if you fight more than one at once.

Roundstart antags are being removed; all the midrounds are usually always pretty numerous (pirates, spiders, blob, buffed aliens) to lay down a good amount of suppressive fire (or, well, "damage", or whatever'd be the alternative for melee lads) or have an equal amount of gun (NRI raiders if people play them the pirate way, assops, whatever else I forgot) compared to what the station can get to do the same but in lesser numbers;
> All even fights become a contest of "who decided to click first / has lower ping after the ritual handshake of Combat Starts Now"

Always has been really.
> Takes all the guns with gimmicky traits and cranks the power of them to 11.

**Insanely good!** 
> Hitscan buckshot

**Insanely good!**
> Hitscan ebows

I forgor brb.
 
> Hitscan accelerator lasers

I forgor brb; wanted not to but iirc they're kinda meant to be a projectile.

**Comment by theselfish** [comment]
Also; Wouldn't this make Stormtrooper Aim (even more) useless?

**Comment by Stalkeros2** [comment]
> Also; Wouldn't this make Stormtrooper Aim (even more) useless?

Don't think so; to be frank, it'll probably get "better" at what it's doing, that is, making your shots unpredictable. As now you won't even know their trajectory after they exit the barrel.

**Comment by Iamgoofball** [comment]
>Insanely good!

this is not, in fact, """insanely good"""

this is taking the most overpowered weapon in the game(buckshot shotguns) and making them ten times more overpowered for, what I can only describe as "lol so random FTW" logic

please deliver a thesis statement and some paragraphs as to why you feel that hitscan is going to be good for the environment of skyrat because all I see is some incorrect claim that it "evens" the playing field(it doesn't)

>With the sheer amount of ranged buffs and Skyrat's very ranged-centric combat, to imply it wasn't already is kinda funny.

a boat leaking is not a reason to put more holes in the hull

**Comment by masline147** [comment]
Hitscan is not very advisable to enter into the code, as the weapon will become powerful. If you make bullets like on TGCM, it will already be better. Ene is not required for lasers, since now lasers are already the best of weapons.

**Comment by SpaceVampire** [comment]
This sounds like a great way to increase the RP level on the server.
So often when you point a gun at people they don't worry about it, because they can just run away.
You also have people darting around trying to dodge the bullets like something out of the Matrix (1999).
Making guns more scary should make people to treat them with more respect.
It is also how guns actually work in real life, so it gets points for that.
Hopefully it will also make range combat (based) more popular then melee combat which is bad.
Giving disablers an actual use compared to one-tap win stun batons.

**Comment by burgerenergy** [comment]
I feel like this will have the opposite effect regarding player skill levels, too. Hitscan over a traveling bolt means you need to actually _click_ on the person squirreling around in combat, whereas with a traveling bolt you have considerable leeway in landing hits, especially considering most map design lends itself to cramped quarters and, at most, three width hallway lanes. Increase the velocity if you want,

**Comment by tf-4** [comment]
Basically what goof said

**Comment by Stalkeros2** [comment]
https://tenor.com/view/cat-death-funny-cute-awesome-gif-20320653

**Comment by Stalkeros2** [comment]
Seething and coping right now.

**Pull request closed without merging**